### DISCUSSION TITLE
Suggested new option for mspass_func_wrapper

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#224] [Design & Development]
I find this in lines 57-63 of the copy I have of decorators.py:
```
    try:
        res = func(data, *args, **kwargs)
        if object_history:
            logging_helper.info(data, alg_id, alg_name)
        if res is None and inplace_return:
            return data
        return res
```
I suggest strongly we remove the test for "res is None".   There is no need to put that restriction on the return of func.   If the function returns something like an int this will fail for no good reason.  I suggest we first drop that test for None.  We should then add an option to post the value of res to Metadata with a specified key.   Seems to me a rational addition would be this arg:  `function_return_key=None`.   The wrapper should then test if function_return_key is None.  When so ignore res, but if so post res to `data` using `function_return_key`.   

What do you think?

**Comment by wangyinz** [comment]
I agree. This is clearly a bug, I also think there are some other issues. We need to have a comprehensive test against this module when implementing the new decorated functions.

**Comment by pavlis** [comment]
I implemented the `function_return_key` idea in `mspass_func_wrapper`, but it raises a question I'm having trouble finding in internet searches.   We already have a bunch of algorithms that use that wrapper. Do I have to modify the arg list of all of them to include the new `function_return_key` argument?   None of them need it so I hope not.    e.g. in signal.py would I need to change:
```
def filter(data, type, *args, object_history=False, alg_name=None, alg_id=None, dryrun=False, inplace_return=True, **options):
```
to
```
def filter(data, type, *args, object_history=False, alg_name=None, alg_id=None, dryrun=False, inplace_return=True, function_return_key=None,**options):
```
I certainly hope that change isn't necessary as it will require a bunch of annoying changes that do nothing.   I'm going to blunder forward assuming this will not be a problem unless you tell me otherwise.  pytest should quickly tell me if this is broken anyway right?

**Reply by pavlis** [reply]
I just noticed another related thing.  For the example above using the decorator for `filter`, why isn't this the arg definition instead of the current one:
```
def filter(data, type, *args, object_history=False, alg_name='filter', alg_id=None, dryrun=False, inplace_return=True, **options):
```
Noting the key point being setting `alg_name='filter'`?

**Reply by wangyinz** [reply]
To my understanding, you will need to add that new argument to all the decorated functions to make the api consistent. Otherwise, the `function_return_key=None` will be passed to the underlying function call as part of the `**options` if that is ever added to the argument list by the user. Might want to confirm with @haruming on this, but this is my understanding.

I am not sure what exactly you are asking in the second questions. Maybe you are confused why the `alg_name` is defined in the argument list here. `alg_name='filter'` is necessary here, because we are **using** the decorator here, and the decorator needs to know the `alg_name` to properly record the history.

**Reply by haruming** [reply]
I agree with you, to make the API consistent. However, I do have doubts on whether it would cause a problem if we don't provide the `function_return_key=None` in the algorithm function using that decorator because the decorator will treat it as None by default so that it will not be passed to the underlying function call as part of the `**options`. I think `**options` are those key-value parameters not specified by the decorator.
Anyway, I will test it to see if that agrees with what I said. Sometimes the functionalities in the `decorator` library are not consistent with what I thought before from my previous experience.

**Reply by wangyinz** [reply]
Good point! Let's test it out first.

**Reply by haruming** [reply]
I test it out and it agrees with what I commented. We don't need to put `function_return_key=None` in any algorithm unless it needs that arg. Since we specify that arg in our mspass_func_wrapper, the arg will not be a part of `**kwargs` in our algorithm function call. In short, any algorithm using the wrapper but not specifying the `function_return_key=None` arg will use the default value(maybe None?) in the wrapper, which means, it will never post the res to the metadata.

**Reply by pavlis** [reply]
Sorry about that confusion.   The point is the current version of our wrapper for filter is this:
```
def filter(data, type, *args, object_history=False, alg_name=None, alg_id=None, dryrun=False, inplace_return=True, **options):
```
I was saying I think that is mistake and we should change it to have `alg_name='filter'`.  From what was said it seems that is a bug  or a least a relatively harmless mistake.

**Reply by pavlis** [reply]
PS to previous:   my local copy had not refreshed I thought this was being inserted after @wangyinz "I'm not sure what exactly you are asking...".   I think I can take this from here, thanks.   I am working on testing what I've done.  I had to do some minor surgery to the C++ code because of a decision I made to add a feature that was missing from the older version of "scale".  Scale had to be modified anyway to use the decorators so I took it a step further.

**Reply by wangyinz** [reply]
Yeah, now I see it... It is clearly a mistake... I should've been more careful reviewing the code.

**Comment by pavlis** [comment]
Yet another wrinkle on how to use these decorators.  In the file mspasspy/algorithms/window.py is a python class I created a while back called `TopMute`.  It has an `apply` method that is a type example of an `inplace_return` in the decorators.   I don't think I can use the mspass_func_wrapper in that context can I?   I'm not going to mess with that function right now as it isn't critical for our short course, but I think we may need to eventually deal with that issue.

**Reply by wangyinz** [reply]
I think we can still use the decorator. Actually, it can simplify these algorithm definition, and we only need to handle the history inside the decorator. I think we need to revise all the C++ functions to have a unified way of handling history and error logs through decorators, but yes, it is not as important at the moment.
